src/controllers/hid/hidioglobaloutputreportfifo.cpp

```diff
@@ -0,0 +1,97 @@
+#include "controllers/hid/hidioglobaloutputreportfifo.h"
+
+#include <hidapi.h>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "util/compatibility/qbytearray.h"
+#include "util/compatibility/qmutex.h"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kMaxHidErrorMessageSize = 512;
+constexpr int kSizeOfFifoInReports = 32;
+} // namespace
+
+HidIoGlobalOutputReportFifo::HidIoGlobalOutputReportFifo()
+        : m_fifoQueue(kSizeOfFifoInReports) {
+}
+
+void HidIoGlobalOutputReportFifo::addReportDatasetToFifo(const quint8 reportId,
+        const QByteArray& reportData,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    // First byte must always be the ReportID-Byte
+    QByteArray report(std::move(reportData));
+    report.prepend(reportId);
+
+    // Swap report to lockless FIFO queue
+    bool success = m_fifoQueue.try_emplace(std::move(report));
+
+    // Handle the case, that the FIFO queue is full - which is an error case
+    if (!success) {
+        // If the FIFO is full, we skip the report dataset even
+        // in non-skipping mode, to keep the controller mapping thread
+        // responsive for InputReports from the controller.
+        // Alternative would be to block processing of the controller
+        // mapping thread, until the FIFO has space again.
+        qCWarning(logOutput)
+                << "FIFO overflow: Unable to add OutputReport " << reportId
+                << "to the global cache for non-skipping sending of OututReports for"
+                << deviceInfo.formatName();
+    }
+}
+
+bool HidIoGlobalOutputReportFifo::sendNextReportDataset(QMutex* pHidDeviceAndPollMutex,
+        hid_device* pHidDevice,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    auto startOfHidWrite = mixxx::Time::elapsed();
+
+    auto pFront = m_fifoQueue.front();
+
+    if (pFront == nullptr) {
+        // No data in FIFO to be send
+        // Return with false, to signal the caller, that no time consuming IO
+        // operation was ncessary
+        return false;
+    }
+
+    auto fifoUsedEntries = m_fifoQueue.size();
```

Comment:
The result of this (costly) operation is used for logging and should not be invoked otherwise. It accesses 2 atomics with acquire ordering and requires a memory fence.

Don't use lock-free data structures like any other, simple container class.